### PR TITLE
[Feature](bangc-ops): add ms_deform_attn_forward to r0.4 branch

### PR DESCRIPTION
**Pull request by DanieeelLiu**
Thanks for your contribution and we appreciate it a lot. :rocket::rocket:

## 1. Motivation

Please describe your motivation and the goal you want to achieve through this pull request.

## 2. Modification

Please briefly describe what modification is made in this pull request, and indicate where to make the modification.

Are new test cases added? If so, please post the corresponding generator-PR link here.

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS™ Accuracy Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

- static threshold
  - diff1
    - [ ] float32 mlu diff1 <= 1e-5
    - [ ] float32 mlu diff1 <= 3e-3
    - [ ] float16 mlu diff1 <= 3e-3
  - diff2
    - [ ] float32 mlu diff2 <= 1e-5
    - [ ] float32 mlu diff2 <= 3e-3
    - [ ] float16 mlu diff2 <= 3e-3
  - diff3
    - [ ] mlu diff3 == 0
    - [ ] mlu diff3_1 == 0
    - [ ] mlu diff3_2 == 0
- dynamic threshold
  - [ ] diff1: mlu diff1 <= max(baseline diff1 * 10, static threshold)
  - [ ] diff2: mlu diff2 <= max(baseline diff2 * 10, static threshold)
  - [ ] diff3: mlu diff3 <= max(baseline diff3 * 10, static threshold)
    - float32, threshold = 1e-5
    - float16, threshold = 1e-3

#### 3.1.2 Operator Scheme checklist

- Supported hardware
  - [ ] MLU370
  - [ ] MLU590
- Job types
  - [ ] BLOCK
  - [ ] UNION1
  - [ ] UNION2
  - [ ] UNION4
  - [ ] The operator will dynamically select the most suitable task type, for example, UNION8

#### 3.2.1 Accuracy Test

If you have checked the following items, please tick the relevant box.

- [ ] Data type test (e.g. float32/int8)
- [ ] Multi-dimensional tensor test
- [ ] Layout test
- [ ] Different size/integer remainder end segment/alignment misalignment test
- [ ] Zero dimensional tensor test/zero element test
- [ ] stability test
- [ ] Multiple platform test
- [ ] Gen_case module test, see: [Gencase-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/Gencase-User-Guide-zh.md)
- [ ] Nan/INF tests 
- [ ] Bug fix tests
- [ ] For memory leak check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md)
- [ ] For code coverage check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md)
- [ ] For I/O calculation efficiency check details, see: [MLU-OPS™-Performance-Acceptance-Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md)

#### 3.2.2 Parameter Check

Test Point-1: `When a new operator is submitted, the test points are given and the test results are stated`. Acceptance Standard: `Normal error`.
```bash
Please fill your test results(Error Message) in here, ...
```

Test Point-2: `Whether illegal parameters are passed`. Acceptance Standard: `Normal error`.
```bash
Test results...
```

### 3.3 Performance Test

See [MLU-OPS™ Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform：MLU370

```bash
# The test results should contain Op name, Shape, Data type,  
#   MLU Hardware Time(us), MLU Interface Time(us), MLU IO Efficiency, 
#   MLU Compute Efficiency, and Mlu Workspace Size(Bytes)
# 
# for example:
#
# ----------- case0 -----------
# case0
# [Op name                ]: abs
# [Shape                  ]: input.shape=[1024,1024,3,4], output.shape=[1024,1024,3,4]
# [Data type]             ]: float32
# [MLU Hardware Time      ]: 15728 (us)
# [MLU Interface Time     ]: 369.008 (us)
# [MLU IO Efficiency      ]: 0.23275
# [MLU Compute Efficiency ]: 0.5
# [Mlu Workspace Size     ]: -1 (Bytes)
# 
# ----------- case1 -----------
# ...
```

Platform：MLU590
```bash
# ----------- case0 -----------
# ----------- case1 -----------
# ...
```

### 3.4 Summary Analysis

mmcv test passed and all case passed.

### CODE REVIEW

**Comment by PetrelYy** [comment]
LGTM

**Review by PetrelYy — APPROVED** [review]

File: bangc-ops/kernels/ms_deform_attn_forward/ms_deform_attn_forward.h [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+#include "mlu_op.h"
+#include "kernels/kernel.h"
+#include "kernels/utils/common.h"
+
+#define MIN(a, b) (((a) < (b)) ? (a) : (b))
+#define MS_DEFORM_ATTN_FORWARD_HEADVECTOR 1
+
+template <typename T>
+__mlu_global__ void MLUKernelMsDeformAttnForwardDefault(
+    const char *data_value_gdram,
```
Comment (chqy99) [inline bangc-ops/kernels/ms_deform_attn_forward/ms_deform_attn_forward.h:35]: char 要替换为 int8_t

File: bangc-ops/kernels/ms_deform_attn_forward/ms_deform_attn_forward.h [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    const char *data_sampling_loc_gdram,
+    const char *data_attn_weight_gdram,
+    const int batch_size,
+    const int num_keys,
+    const int num_heads,
+    const int channels,
+    const int num_levels,
+    const int num_queries,
+    const int num_points,
+    char *data_col_gdram);
```
Comment (chqy99) [inline bangc-ops/kernels/ms_deform_attn_forward/ms_deform_attn_forward.h:47]: 同理

File: bangc-ops/kernels/ms_deform_attn_forward/ms_deform_attn_forward.mlu [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *************************************************************************/
+#include "kernels/ms_deform_attn_forward/ms_deform_attn_forward.h"
+
+#include "core/context.h"
+#include "core/logging.h"
```
Comment (chqy99) [inline bangc-ops/kernels/ms_deform_attn_forward/ms_deform_attn_forward.mlu:26]: 字母顺序，g应该在l之前

File: bangc-ops/kernels/ms_deform_attn_forward/ms_deform_attn_forward.mlu [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+#include "core/gen_case.h"
+#include "core/runtime/device.h"
+#include "core/tensor.h"
+#include "core/type.h"
```
Comment (chqy99) [inline bangc-ops/kernels/ms_deform_attn_forward/ms_deform_attn_forward.mlu:30]: tool 在 type 之前

File: bangc-ops/kernels/ms_deform_attn_forward/ms_deform_attn_forward.mlu [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+  const int32_t num_points = data_sampling_loc_desc->dims[4];
+  // check element num zero
+  if (batch_size == 0 || num_heads == 0 || channels == 0 || num_queries == 0) {
+    LOG(ERROR) << "[mluOpMsDeformAttnForward] Check failed: element num zero.";
+    return MLUOP_STATUS_BAD_PARAM;
+  }
+  if (num_levels == 0 || num_points == 0) {
+    VLOG(5) << "mluOpFill_v3 start.";
+    const float fill_value = 0.0f;
+    PARAM_CHECK("[mluOpMsDeformAttnForward]",
```
Comment (chqy99) [inline bangc-ops/kernels/ms_deform_attn_forward/ms_deform_attn_forward.mlu:195]: PARAM_CHECK 旨在检查参数，函数调用应该用 CHECK_RETURN

File: bangc-ops/kernels/ms_deform_attn_forward/ms_deform_attn_forward.mlu [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+          VLOG(5) << "Not Implemented";
+          break;
+        }
+        case CNRT_FUNC_TYPE_BLOCK: {
+          VLOG(5)
+              << "Launch Kernel MLUKernelMsDeformAttnForwardDefault<<<Block, "
+              << k_dims.x << ", " << k_dims.y << ", " << k_dims.z << ">>>";
+          KERNEL_CHECK((MLUKernelMsDeformAttnForwardDefault<
+                        float><<<k_dims, k_type, handle->queue>>>(
+              (char *)data_value, (char *)data_spatial_shapes,
```
Comment (chqy99) [inline bangc-ops/kernels/ms_deform_attn_forward/ms_deform_attn_forward.mlu:254]: 同理，char *类型应替换为 int8_t

File: bangc-ops/kernels/ms_deform_attn_forward/ms_deform_attn_forward.h (hunk 1 above)
Comment (DanieeelLiu) [reply bangc-ops/kernels/ms_deform_attn_forward/ms_deform_attn_forward.h:35]: 这是r0.4 分支的，toolkit是 3.0.2

File: bangc-ops/kernels/ms_deform_attn_forward/ms_deform_attn_forward.h (hunk 2 above)
Comment (DanieeelLiu) [reply bangc-ops/kernels/ms_deform_attn_forward/ms_deform_attn_forward.h:47]: 同上

File: bangc-ops/kernels/ms_deform_attn_forward/ms_deform_attn_forward.mlu (hunk 6 above)
Comment (DanieeelLiu) [reply bangc-ops/kernels/ms_deform_attn_forward/ms_deform_attn_forward.mlu:254]: 同上

**Comment by chqy99** [comment]
LGTM

**Review by chqy99 — APPROVED** [review]

**Review by duzekunKTH — APPROVED** [review]
LGTM

**Pull request merged**